// Module/Module.js
var fs = require('fs-promise')
var filepath = require('path')

var style = `
:host {
  
}
::content textarea { height: 15rem; }
::content li { margin-bottom: 1rem; }
`

module.exports = class Module extends WebComponentAbstract {
  init() {
    this.renderCss(style)
    
    this.userModulesDir = filepath.join(require('electron').remote.app.getPath('userData'), 'node_modules')
    
    var div = this.new('div', { className: 'column-system' })
    var left = div.new('div')
    var right = div.new('div')
    
    var p = left.new('p')
    p.newText('Each line of module list represents a module name.')
    p.new('br')
    p.newText('Be careful load only trusted modules.')
    
    this.ser.conf = this.ser.conf || ''
    Synth(this.ser, 'conf', left.new('p').new('textarea', { placeholder: 'Module names' }))
    left.new('p').innerHTML = `<a tabIndex="0" onclick="location.reload()" class="button">Reload to take effect</a>`
    
    this.loadUserModulesDir().then(e => this.exampleModule().then(e => {
      this.loadModules()
      
      var exampleFile = moduleAvailable('Example')
      right.new('h1').textContent = 'Modules development'
      right.new('div').innerHTML = `
<ol>
  <li><a tabIndex="0" onclick="require('electron').shell.showItemInFolder(${JSON.stringify(exampleFile).replace(/"/g, '&quot;')})" title="${exampleFile}" class="button">Take Example.js</a></li>
  <li>Put some code in it</li>
  <li><a tabIndex="0" onclick="self.emit('newModule', 'Example')" class="button">Add it to the module list</a></li>
  <li><a tabIndex="0" onclick="location.reload()" class="button">Reload app</a></li>
  <li><a tabIndex="0" onclick="require('electron').remote.getCurrentWindow().toggleDevTools()" class="button">Toggle developer tools for debug</a></li>
  <li><a tabIndex="0" href="${app.package.repository.url}" class="button">Make a pull request</a></li>
  <li>Spread the word!</li>
</ol>
      `
    }))
    
    var p = this.new('p', { textContent: 'Open your require path: ' + this.userModulesDir, tabIndex: 0 })
    p.on('click', e => require('electron').shell.showItemInFolder(this.userModulesDir))
    p.style.cursor = 'pointer'
    
    defineEvent('newModule', 'Adds new module to module list', 'Core', `''`)
    self.on('newModule', e => this.ser.conf += e.detail + '\n')
    
    self.emit('newTab', { el: this, name: 'Modules', priority: 850 })
  }
  
  loadUserModulesDir() {
    module.paths.unshift(this.userModulesDir)
    return fs.ensureDir(this.userModulesDir + '/Example')
  }
  
  exampleModule() {
    return new Promise(resolve => {
      fs.stat(this.userModulesDir + '/Example/package.json', e => {
        if(!e) { return resolve() }
        
        Promise.all([
          fs.copy(__dirname + '/Example/package.json', this.userModulesDir + '/Example/package.json'),
          fs.copy(__dirname + '/Example/Example.js', this.userModulesDir + '/Example/Example.js'),
        ]).then(resolve)
      })
    })
  }
  
  loadModules() {
    this.ser.conf.split('\n').forEach(s => s && moduleAvailable(s) && setImmediate(() => app.new(require(s), false)))
  }
}

var moduleAvailable = m => { try { return require.resolve(m) } catch(e) { return false } }
