// Serial/Buffer.js
module.exports = class SerialBuffer extends WebComponentAbstract {
  init() {
    this.new(require('./StartCheck'))
    this.new(require('./Ready'))
    
    this.buffer = []
    
    defineEvent('serialWrite', 'Send data to serial buffer', 'Serial buffer', `{ data: '' or [], prepend: false }`)
    self.on('serialWrite', e => this.send(e.detail.data, e.detail.prepend))
    
    defineEvent('serialClean', 'Clean serial buffer', 'Serial buffer')
    self.on('serialClean', e => { this.buffer = []; self.emit('bufferChange'); })
    
    defineEvent('deviceOk', 'Device ok, listen only', 'Serial buffer')
    defineEvent('deviceNotOk', 'Device busy, listen only', 'Serial buffer')
    this.deviceOk = false
    self.on('deviceOk', e => this.deviceOk = true)
    self.on('deviceNotOk', e => this.deviceOk = false)
    self.on('disconnected', e => self.emit('deviceNotOk'))
    
    self.on('deviceReady', e => self.emit('deviceOk'))
    self.on('serialData', e => this.checkOk(e))
    self.on('deviceOk', e => setImmediate(e => this.sendBuffer()))
    self.on('deviceReset', e => self.emit('deviceReady'))
    
    setImmediate(() => self.emit('bufferChange'))
  }
  
  checkOk(e) {
    if(!this.deviceReady) return
    if(e.detail.data.substr(0, 2) == 'ok') self.emit('deviceOk')
  }
  
  send(data, prepend) {
    if(!Array.isArray(data)) data = String(data).split('\n')
    
    if(prepend) data.reverse()
    
    data.forEach(i => {
      if(!i) return
      prepend ? this.buffer.unshift(i + '\n') : this.buffer.push(i + '\n')
    })
    
    self.emit('bufferChange')
    this.sendBuffer()
  }
  
  sendBuffer() {
    if(!this.deviceOk) return
    
    var data = this.buffer.shift()
    if(!data) return
    
    self.emit('bufferChange')
    self.emit('deviceNotOk')
    self.emit('serialDirectWrite', { data })
  }
}
