// Module/Example/Example.js
// less stylesheet for our module
var style = `
:host {
  
}

::content {
  button {
    display: block;
    height: 100%;
    width: 100%;
    font-size: 2em;
    color: #999;
    background-image: repeating-linear-gradient(
      -45deg,
      lighten(#000, 100%),
      lighten(#000, 100%) 9px,
      lighten(#000, 94%) 10px,
      lighten(#000, 97%) 11px,
      lighten(#000, 97%) 20px
    );
  }
}
`

// modules are custom html elements
module.exports = class NameOfYourModule extends WebComponentAbstract {
  init() {
    // render stylesheet above
    this.renderLess(style)
    
    // create new <button>
    this.but = this.new('button')
    
    // some text in it
    this.but.textContent = 'Emergency stop'
    
    // on click we will dispatch app event serialWrite with some gcode
    // all events are documented in Developer tab
    this.but.on('click', e => self.emit('serialWrite', { data: 'M112', prepend: true }))
    
    // dispatch events, in this case we are going for new tab
    self.emit('newTab', { el: this, 'name': 'Example' })
  }
}
