// Control/UI/AxisSettings.js
var style = `
::content label { white-space: nowrap; }
::content input[type='number'] { width: 5rem; }
`

module.exports = class AxisSettings extends WebComponentAbstract {
  init() {
    this.renderCss(style)
    
    var p = this.new('p', { innerHTML: this.opt.name })
    p.new('br')
    
    var opt = { type: 'number', step: 1 }
    this.X = p.new('label', { textContent: 'X '}).new('input', opt)
    this.X.on('input', e => this.set())
    p.newText(' ')
    
    this.Y = p.new('label', { textContent: 'Y '}).new('input', opt)
    this.Y.on('input', e => this.set())
    p.new('br')
    
    this.Z = p.new('label', { textContent: 'Z '}).new('input', opt)
    this.Z.on('input', e => this.set())
    p.newText(' ')
    
    this.E = p.new('label', { textContent: 'E '}).new('input', opt)
    this.E.on('input', e => this.set())
    p.new('br')
    
    self.on('serialConfig', e => this.echo(e.detail))
  }
  
  set() {
    var val = `X${this.X.value} Y${this.Y.value} Z${this.Z.value} E${this.E.value}`
    self.emit('serialWrite', { data: `${this.opt.gCode} ${val}`, prepend: true })
  }
  
  value(msg) {
    var m = msg.match(/([0-9]+\.?[0-9]*)/g)
    this.X.value = parseFloat(m[1])
    this.Y.value = parseFloat(m[2])
    this.Z.value = parseFloat(m[3])
    this.E.value = parseFloat(m[4])
  }
  
  echo(msg) {
    if(msg.startsWith(this.opt.gCode)) return this.value(msg)
  }
}
