// Developer/Developer.js
var style = `
:host {
  display: block;
  height: 100%;
  overflow-x: hidden;
  overflow-y: auto;
}

::content {
  
}
`
var info = `
<h1>Globals</h1>
<p>
  Var App is pointer to &lt;app-pracontrol&gt; with extra props .ready and .package.<br>
  All functions are defined in web-component-app module.
</p>
`

module.exports = class Developer extends WebComponentAbstract {
  
  init() {
    this.opened = false
    
    defineEvent('developerMode', 'Turns on developer mode and this page', 'Developer')
    self.on('developerMode', e => !this.opened && this.turnOn())
    
    this.new('div').innerHTML = info
    this.new(require('./AppEventTable.js'))
  }
  
  turnOn() {
    this.opened = true
    
    this.renderCss(style)
    
    AppEvent('newTab', { elem: this, name: 'Developer', priority: 900 })
    
    defineShortcut('⌥', 'Toogle baseline grid', 'Developer')
    document.on('keydown', e => e.keyCode == 18 && document.body.classList.toggle('alt_key'))
    document.on('keyup', e => e.keyCode == 18 && document.body.classList.toggle('alt_key'))
    document.body.new('rembased_grid')
  }
}
