// Tab/Tab.js
var style = `
:host {
  order: -100;
  flex-grow: 1;
  display: flex;
  flex-wrap: wrap;
  
  background-color: #333;
}

::content {
  
  > tab {
    padding: 0.5rem;
    color: #fff;
    font-size: 0.85em;
    
    &.active, &:hover { background-color: #666; }
    
    &[name='PraControl'] {
      background-image: url('res/home.svg');
      background-position: center center;
      background-repeat: no-repeat;
      background-size: 60%;
      font-size: 0;
      width: 1rem;
    }
  }
}
`

module.exports = class Tab extends TabComponent {
  init() {
    this.renderLess(style)
    
    this.initTab()
    self.on('newTab', e => { this.newTab(e.d); AppEvent('newPanel', { elem: e.d.elem })})
    self.on('focusTab', e => this.focusTab(e.d))
    
    defineShortcut('⌘1‑9', 'Go to tab 1‑9', 'Tabs')
    self.on('keydown', e => (e.metaKey || e.ctrlKey) && this.tabNumberShortcut(e))
  }
  
  readyCallback() {
    AppEvent('newPanel', { elem: this, position: 'top' })
  }
  
  tabNumberShortcut(e) {
    if(e.keyCode < 49 || e.keyCode > 57) return
    this.focusTab(e.keyCode - 49)
  }
}
