// KeyShortcutsTable/KeyShortcutsTable.js
var style = `
:host {
  display: block;
  height: 100%;
  overflow-x: hidden;
  overflow-y: auto;
}
`

module.exports = class KeyShortcutsTable extends WebComponentAbstract {
  init() {
    this.renderCss(style)
    
    defineAppEvent('defineKeyShortcut', 'Add new item to shortcut table', 'Core', "{ key: '', desc: '' }")
    self.on('defineKeyShortcut', e => this.newShortcut(e.d))
    
    this.el = this.newElem(FlexBalanced)
  }
  
  readyCallback() {
    AppEvent('newTab', { elem: this, name: 'Shortcuts', priority: 800 })
  }
  
  newShortcut(k) {
    var div = this.el.querySelector(`div[cat='${k.cat}']`)
    
    if(!div) {
      div = this.el.newElem('div')
      div.setAttribute('cat', k.cat)
      var table = div.newElem('table')
      var thead = table.newElem('thead')
      
      var th = thead.newElem('tr').newElem('th', true, { colSpan: 2, textContent: k.cat })
      
      table.newElem('tbody')
    }
    
    var tr = div.querySelector('tbody').newElem('tr')
    tr.setAttribute('cat', k.cat)
    tr.newElem('td').newElem('kbd').textContent = k.key
    tr.newElem('td').textContent = k.desc
  }
}
