// KeyShortcutsTable/KeyShortcutsTable.js
var style = `
:host {
  display: block;
  height: 100%;
  overflow-x: hidden;
  overflow-y: auto;
}
`

module.exports = class KeyShortcutsTable extends WebComponentAbstract {
  init() {
    this.renderCss(style)
    
    defineEvent('defineShortcut', 'Add new item to shortcut table', 'Core', "{ key: '', desc: '' }")
    self.on('defineShortcut', e => this.newShortcut(e.detail))
    
    this.el = this.new(FlexBalanced)
    
    setImmediate(() => self.emit('newTab', { el: this, name: 'Shortcuts', priority: 800 }))
  }
  
  newShortcut(k) {
    var div = this.el.querySelector(`div[cat='${k.cat}']`)
    
    if(!div) {
      div = this.el.new('div')
      div.setAttribute('cat', k.cat)
      var table = div.new('table')
      var thead = table.new('thead')
      
      var th = thead.new('tr').new('th', { colSpan: 2, textContent: k.cat })
      
      table.new('tbody')
    }
    
    var tr = div.querySelector('tbody').new('tr')
    tr.setAttribute('cat', k.cat)
    tr.new('td').new('kbd').textContent = k.key
    tr.new('td').textContent = k.desc
  }
}
