// Printer/Printer.js
var fs = require('fs')

var style = `
:host {
  
}

::content {
  drop {
    display: block;
    height: 10rem;
    width: 100%;
    border-radius: 1rem;
    display: flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    border: 1px dashed LightGray;
    :host-context(.drag) & { background-color: WhiteSmoke; }
  }
}
`

module.exports = class Printer extends WebComponentAbstract {
  init() {
    this.renderLess(style)
    
    var drop = this.new('drop', { textContent: 'Select or drop .gcode for print', tabIndex: 0 })
    drop.on('click', e => self.emit('openFileDialog'))
    drop.style.cursor = 'pointer'
    
    self.on('openFile', e => this.openFile(e))
    
    self.emit('newControl', { el: this })
  }
  
  openFile(e) {
    e.detail.forEach(f => f.path.substr(-5) == 'gcode' && this.printFile(f.path))
  }
  
  printFile(path) {
    fs.readFile(path, function(e, data) {
      if(e) { throw e }
      self.emit('serialWrite', { data: data.toString() })
    })
  }
}
