// Printer/Printer.js
var fs = require('fs');

var style = `
:host {
  
}

::content {
  drop {
    display: block;
    height: 10rem;
    width: 100%;
    border-radius: 1rem;
    display: flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    border: 1px dashed LightGray;
    :host-context(.drag) & { background-color: WhiteSmoke; }
  }
}
`

module.exports = class Printer extends WebComponentAbstract {
  init() {
    this.renderLess(style)
    
    this.drop = this.new('drop', true, { textContent: 'Drop .gcode for print' })
    
    self.on('newFile', e => this.newFile(e))
    
    setImmediate(() => self.emit('newControl', { elem: this }))
  }
  
  newFile(e) {
    e.detail.forEach(file => {
      if(file.path.substr(-5) == 'gcode') this.printFile(file.path)
    })
  }
  
  printFile(path) {
    fs.readFile(path, function(err, data) {
        if(err) throw err;
        
        var file = data.toString().split('\n');
        self.emit('serialWrite', { data: file })
    })
  }
}
