// About/About.js
var style = `
::host {
  
}

::content {
  > .column-system {
    height: 100%;
    align-items: center;
    > left { text-align: center; }
  }
  
  .logo {
    background: url('resources/logo.svg') center center no-repeat;
    background-size: 90% 90%;
    display: block;
    height: 10rem;
    margin-bottom: 1rem;
  }
}
`

var html = `
<div class="column-system">

<left class="column-width-gold">
  <div class="logo"></div>
  <h1>${app.package.productName} <small>${app.package.version}</small></h1>
  <p>${app.package.description}</p>
  <p><a href="${app.package.homepage}">website</a> | <a href="${app.package.repository.url}">GitHub</a></p>
  <p>Contributors: ${app.package.contributors.map(e => `<a href="${e.url || 'mailto://' + e.email}">${e.name}</a>`).join(', ')}</p>
  <p><a href="${app.package.donate}" class="button">Buy a beer 🍺</a></p>
</div>
`

module.exports = class About extends WebComponentAbstract {
  init() {
    this.innerHTML = html
    this.renderLess(style)
    
    self.emit('newTab', { el: this, name: 'About', priority: 1000 })
  }
}
